### PR TITLE
feat(mobile): preserve mobile album info on upload

### PR DESCRIPTION
**Pull request by alextran1502**
This PR brings the ability to sync the selected album for backup with the Immich server. If an asset is in one of more selected albums, for example, a video in `Videos` and `Recents` album, it will be added to both. 

When this option is toggled, the selected albums and any additional albums selected for backup are created on the server with the same name.

There is also an option to sync uploaded photos to the albums to cover the current Immich user. It can also be used to keep everything in sync if some edge cases happen and some uploaded assets don't make it to the album.

| Sync toggle option | Manual sync button |
| - | - |
|  | Screenshot_1724606496 |

### CODE REVIEW

File: mobile/lib/providers/backup/manual_upload.provider.dart [hunk 1]
```diff
@@ -209,7 +207,9 @@ class ManualUploadNotifier extends StateNotifier<ManualUploadState> {
           );
         }
 
-        Set<AssetEntity> allUploadAssets = allAssetsFromDevice.nonNulls.toSet();
+        Set<BackupCandidate> allUploadAssets = allAssetsFromDevice.nonNulls
+            .map((asset) => BackupCandidate(asset: asset, albumNames: []))
```
Comment (shenlong-tanwen) [inline mobile/lib/providers/backup/manual_upload.provider.dart:211]: Does this mean that manual backups do not preserve the album info even when the setting is turned on? It would be best if uploads, from foreground / background / manual, all respect the sync settings

File: mobile/lib/services/album.service.dart [hunk 2]
```diff
@@ -441,4 +444,34 @@ class AlbumService {
...
   }
+
+  Future<Album?> getAlbumByName(String name, bool remoteOnly) async {
+    final query = _db.albums.filter().nameEqualTo(name).sharedEqualTo(false);
+
+    if (remoteOnly) {
+      return query.localIdIsNull().findFirst();
+    }
+
+    return query.findFirst();
```
Comment (shenlong-tanwen) [inline mobile/lib/services/album.service.dart:455]: The query can be constructed in a  single statement using the Isar's [optional](https://isar.dev/queries.html#query-modifiers) method

File: mobile/lib/services/backup.service.dart [hunk 3]
```diff
@@ -81,54 +92,48 @@ class BackupService {
...
     List<AssetPathEntity?> result = [];
-    for (BackupAlbum a in albums) {
+
+    for (BackupAlbum backupAlbum in albums) {
       try {
         final AssetPathEntity album =
             await AssetPathEntity.obtainPathFromProperties(
-          id: a.id,
-          optionGroup: filter.copyWith(
-            updateTimeCond: DateTimeCond(
```
Comment (fyfrey) [inline mobile/lib/services/backup.service.dart:126]: Why is this removed?
This will greatly increase the time of the background upload each time it's executed for large libraries leading to not uploading stuff at all on iOS and wasting battery on Android as it's triggered very often (after any assets changed on device)

File: mobile/lib/services/backup.service.dart [hunk 4]
```diff
@@ -139,48 +144,71 @@ class BackupService {
...
+  Future<Set<BackupCandidate>> _fetchAssetsAndUpdateLastBackup(
     List<AssetPathEntity?> albums,
     List<BackupAlbum> backupAlbums,
-    DateTime now,
   ) async {
-    List<AssetEntity> result = [];
-    for (int i = 0; i < albums.length; i++) {
-      final AssetPathEntity? a = albums[i];
-      if (a != null &&
-          a.lastModified?.isBefore(backupAlbums[i].lastBackup) != true) {
```
Comment (fyfrey) [inline mobile/lib/services/backup.service.dart:151]: Why is this removed?
This will greatly increase the time of the background upload each time it's executed for large libraries leading to not uploading stuff at all on iOS and wasting battery on Android as it's triggered very often (after any assets changed on device)

File: mobile/lib/services/asset.service.dart [hunk 5]
```diff
@@ -284,4 +295,60 @@ class AssetService {
...
+      );
+
+      if (duplicates != null) {
+        candidates
+            .removeWhere((c) => !duplicates.existingIds.contains(c.asset.id));
+      }
+
+      await refreshRemoteAssets();
+      final remoteAssets = await _db.assets
+          .filter()
```
Comment (fyfrey) [inline mobile/lib/services/asset.service.dart:323]: Use where() for one of the two filters (faster because it then uses an index)

File: mobile/lib/services/backup.service.dart (hunk 3 above)
Comment (alextran1502) [reply mobile/lib/services/backup.service.dart:126]: I found that with this filtering condition, and with the scenario of a fresh installed, and choosing let's say "Camera" and "Recent" album for backup. When running in background mode, only the "Recent" album gets processed for the included albums. 

Any thoughts?

File: mobile/lib/pages/backup/backup_album_selection.page.dart [hunk 6]
```diff
@@ -296,15 +254,7 @@ class BackupAlbumSelectionPage extends HookConsumerWidget {
               ],
             ),
           ),
-          SliverLayoutBuilder(
```
[the remark this answers is not in the document]
Comment (alextran1502) [reply mobile/lib/pages/backup/backup_album_selection.page.dart:299]: I can put it back

File: mobile/lib/services/backup.service.dart (hunk 3 above)
Comment (fyfrey) [reply mobile/lib/services/backup.service.dart:126]: I'm not sure, but something in this PR might have changed the logic. It works fine for multiple albums and backs up all assets on main/release.

I believe this requires some debugging (has the album modified date changed: yes/no, are less assets obtained from the album than expected: yes/no, ...).
It's weird that only the background backup is affected - as the code is identical.

File: mobile/lib/services/backup.service.dart (hunk 4 above)
Comment (alextran1502) [reply mobile/lib/services/backup.service.dart:151]: Fixed

File: mobile/lib/services/backup.service.dart (hunk 3 above)
Comment (alextran1502) [reply mobile/lib/services/backup.service.dart:126]: Ah, it was an error in the conditional parameter that was unused. It is now put back and work as expected

File: mobile/lib/providers/backup/manual_upload.provider.dart (hunk 1 above)
Comment (alextran1502) [reply mobile/lib/providers/backup/manual_upload.provider.dart]: Done!

**Review by shenlong-tanwen — APPROVED** [review]
LGTM!

File: mobile/lib/services/album.service.dart [hunk 7]
```diff
@@ -441,4 +444,33 @@ class AlbumService {
       return false;
     }
   }
+
+  Future<Album?> getAlbumByName(String name, bool remoteOnly) async {
+    return _db.albums
+        .filter()
+        .optional(remoteOnly, (q) => q.localIdIsNull())
+        .nameEqualTo(name)
+        .sharedEqualTo(false)
```
Comment (fyfrey) [inline mobile/lib/services/album.service.dart:453]: what if the user shared their automatically-created album (e.g. camera)?

File: mobile/lib/services/asset.service.dart [hunk 8]
```diff
@@ -284,4 +295,61 @@ class AssetService {
       return Future.value(null);
     }
   }
+
+  Future<void> syncUploadedAssetToAlbums() async {
+    try {
+      final selectedAlbums = _backupService.selectedAlbumsQuery().findAllSync();
```
Comment (fyfrey) [inline mobile/lib/services/asset.service.dart:301]: these two lines should use `await` and `findAll()` instead of `findAllSync`

File: mobile/lib/services/album.service.dart (hunk 7 above)
Comment (alextran1502) [reply mobile/lib/services/album.service.dart:453]: The PR's primary goal is syncing the information from the local database to the remote database. For example, if you choose the auto-created `Camera` album. This function will search if there is such album's name existed on the Immich server, if not, then it will create a remote album with the name `Camera` and start adding the uploaded assets to it.

**Comment by pareis** [comment]
Very excited to see this happening!

**Comment by renhiyama** [comment]
@alextran1502 I have already uploaded all my albums from my phone to immich, and even made my parents do too. I don't want to remove the existing uploaded photos and reupload them with preserving the album info. Is it possible to update those previously uploaded images to automatically create new albums on immich server and move those images to their respective albums?

**Comment by alextran1502** [comment]
@renhiyama yes, the 2nd screenshot from the pr description will allow you to do so

**Comment by renhiyama** [comment]
Awesome! Immich cli seems to have this feature already though haha. Thanks for the clarification!

**Comment by Cooper724** [comment]
If I delete a photo on my phones camera roll that I have already synced to immich, will it also delete that photo on immich? Thanks :)

**Comment by alextran1502** [comment]
@Cooper724 no it won't

**Comment by alextran1502** [comment]
@meerumschlungen It will not be removed from the second album. The sync info only happen at the time it is uploaded to the server

**Comment by meerumschlungen** [comment]
> @meerumschlungen It will not be removed from the second album. The sync info only happen at the time it is uploaded to the server

@alextran1502 Thank you for your answer. I naively expected the sync button to retrieve the album information from the camera roll asset again, add the immich asset to albums which it was not a part of before (creating the album in immich if needed) and removing it from immich albums, which it is not part of anymore in the camera roll. If the source asset is not available anymore at all, I would expect the immich asset to be left untouched, preserving the current album "memberships".

Based on your answer, I don't quite understand, what the manual album sync button does (or could ever do), assuming the album info is only fetched once during initial import.

**Comment by alextran1502** [comment]
The manual sync button is for existing user to synchronize the uploaded assets to the albums

It is also not a bad idea to add the feature you want as the enhancement for this mechanism

**Comment by meerumschlungen** [comment]
Just made a few tests. This part

> add the immich asset to albums which it was not a part of before (creating the album in immich if needed)

actually does happen. So there is some mechanism, which extracts the album information even after the initial import (which seemed clear since otherwise the sync option would make no sense).
What am I getting wrong?

**Comment by alextran1502** [comment]
It means only consider adding, not removing assets
